adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -978,20 +978,20 @@ protected void onCommandReceived(final TenantObject tenantObject, final ProtonSe
                     "empty notifications must not contain payload"));
         }
 
-        return contentTypeCheck.future()
-                .compose(ok -> {
-                    switch (context.getEndpoint()) {
-                    case TELEMETRY:
-                        return doUploadMessage(context, resource, getTelemetrySender(resource.getTenantId()), currentSpan);
-                    case EVENT:
-                        return doUploadMessage(context, resource, getEventSender(resource.getTenantId()), currentSpan);
-                    case COMMAND_RESPONSE:
-                        return doUploadCommandResponseMessage(context, resource, currentSpan);
-                    default:
-                        return Future
-                                .failedFuture(new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND, "unknown endpoint"));
-                    }
-                });
+        switch (context.getEndpoint()) {
+        case TELEMETRY:
+            return contentTypeCheck.future()
+                    .compose(ok -> doUploadMessage(context, resource, getTelemetrySender(resource.getTenantId()),
+                            currentSpan));
+        case EVENT:
+            return contentTypeCheck.future()
+                    .compose(ok -> doUploadMessage(context, resource, getEventSender(resource.getTenantId()),
+                            currentSpan));
+        case COMMAND_RESPONSE:
+            return doUploadCommandResponseMessage(context, resource, currentSpan);
+        default:
+            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND, "unknown endpoint"));
+        }
```

Comment:
FMPOV the only thing that needs to be changed/fixed is the *isPayloadOfIndicatedType* method which should verify that the payload is indeed empty if the content type is `EventConstants.CONTENT_TYPE_EMPTY_NOTIFICATION` but otherwise return `true`. WDYT?

Reply:
That makes sense. The problem I see here is that the method does not do what its name promises, and the method is called in all protocol adapters for every telemetry and event message. My suggestion would be to change it to the way you described it. Additionally, we rename it, for example to `isEmtpyNotificationReallyEmpty`. 
Do we also want to change the invocations so that it is only executed for events?

Reply:
> My suggestion would be to change it to the way you described it. Additionally, we rename it, for example to isEmtpyNotificationReallyEmpty.

+1

> Do we also want to change the invocations so that it is only executed for events?

No, let's keep it the way it is. IIRC we allow devices to send empty notifications as both telemetry as well as event messages ...

Reply:
Just to be sure: we are changing the behavior of the protocol adapters by that in that they now generally accept empty messages, aren't we? At least one unit test suggest that it was intended to block empty messages (except the empty notification): https://github.com/eclipse/hono/blob/0840f2e6c173d3d47fba6cbdfecf5ceb363e0147/service-base/src/test/java/org/eclipse/hono/service/AbstractProtocolAdapterBaseTest.java#L483
This test fails with the change discussed here. I've seen no failing IT for this.

Reply:
Yeah, we originally thought it would have value to prevent devices from sending empty messages. However, I can't recall what that value would actually be. On the other hand, I can imagine a device sending an empty message with an application specific content-type which can be used to trigger some behavior in a downstream application. Requiring the device to include a byte in the payload seems useless, doesn't it?
The Telemetry API currently defines that the payload of a message MUST consist of a Data section. The same definition is used by the Event API which doesn't seem to make sense since we already allow the payload to be empty for empty notification (events). The C&C API explicitly allows the payload of a command response to be empty.
So, FMPOV we should
* proceed as discussed with this PR
* remove the conflicting unit test
* adapt the Telemetry API wording to make the payload optional

Reply:
The user guides of the protocol adapters state as well that the payload is required. Should we change the documentation there as well? FMPOV this needs to be tested before changing the user guide. If we remove the explicit check, empty messages might or might not work in the adapters.

Reply:
You could add corresponding unit tests, I guess

Reply:
And yes, I think we should adapt the user guides accordingly as well.

Reply:
The MQTT adapter does not allow an empty payload and explicitly checks that (removing this check involves adding null checks further down...). The HTTP adapter, on the other hand, takes messages with empty payload, but then does not set the content type (`MessageHelper.setPayload()`). 
Of course, this could be changed, but I'm not sure if we want to change the behavior of the protocol adapters in a fix for one specific adapter on the last day before the release. 
Anyway, we should have a separate issue for this.FMPOV allowing empty messages with custom content types is a great idea but a separate feature and not part of this bug fix. WDYT?

Reply:
That is probably true. Do the other adapters allow empty command responses?

Reply:
COAP: no, Lora: no, Sigfox: yes.

Reply:
Well at least the CoAP and Sigfox adapters do not support C&C at all.
The Lora adapter should be fixed as well, I guess. What about the MQTT and HTTP adapters?

Reply:
I'm sorry @sophokles73, I misread your question. My answer referred to which adapters explicitly reject empty payload for messages in general (i.e. especially telemetry and events). The bug in this ticket affects definitively only the AMQP adapter. All other adapters seem to have no problem with empty command responses.

Reply:
ok, then let's fix the AMQP adapter and add unit tests that make sure that all adapters which support C&C accept empty command responses.